cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,858 @@
+                 internal::Assembly::CopyData::CopyDataBase<dim> &data_base) const
+        {
+            internal::Assembly::Scratch::StokesPreconditioner<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::StokesPreconditioner<dim>& > (scratch_base);
+            internal::Assembly::CopyData::StokesPreconditioner<dim> &data = dynamic_cast<internal::Assembly::CopyData::StokesPreconditioner<dim>& > (data_base);
+            //std::cout<<"StokesPreconditioner was called";
+            const MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
+            scratch.material_model_outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >();
+            
+            const Introspection<dim> &introspection = this->introspection();
+            const FiniteElement<dim> &fe = this->get_fe();
+            const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
+            const unsigned int n_q_points           = scratch.finite_element_values.n_quadrature_points;
+            const double pressure_scaling = this->get_pressure_scaling();
+            
+            // First loop over all dofs and find those that are in the Stokes system
+            // save the component (pressure and dim velocities) each belongs to.
+            for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell; /*increment at end of loop*/)
+            {
+                if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+                {
+                    scratch.dof_component_indices[i_stokes] = fe.system_to_component_index(i).first;
+                    ++i_stokes;
+                }
+                ++i;
+            }
+            
+            // Loop over all quadrature points and assemble their contributions to
+            // the preconditioner matrix
+            for (unsigned int q = 0; q < n_q_points; ++q)
+            {
+                for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell; /*increment at end of loop*/)
+                {
+                    if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+                    {
+                        scratch.grads_phi_u[i_stokes] =
+                        scratch.finite_element_values[introspection.extractors
+                                                      .velocities].symmetric_gradient(i, q);
+                        scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection
+                                                                                .extractors.pressure].value(i, q);
+                        ++i_stokes;
+                    }
+                    ++i;
+                }
+                
+                const double eta = scratch.material_model_outputs.viscosities[q];
+                const double one_over_eta = 1. / eta;
+                
+                const bool use_tensor = (anisotropic_viscosity != nullptr);

```

Comment:
Should that not always be the case for you? If so, you could do
```
  Assert (anisotropic_viscosity != null_ptr, ...);
```
and replace `use_tensor` by `true` in everything below.